sql/core/src/test/scala/org/apache/spark/sql/sources/FSBasedRelationSuite.scala

```diff
@@ -0,0 +1,425 @@
+    val data = sparkContext.parallelize(Seq.empty[String])
+    data.saveAsTextFile(new Path(basePath, "p1=1/p2=hello").toString)
+    data.saveAsTextFile(new Path(basePath, "p1=2/p2=world").toString)
+
+    val dataSchema =
+      StructType(
+        Seq(
+          StructField("a", IntegerType, nullable = false),
+          StructField("p1", IntegerType, nullable = true),
+          StructField("b", StringType, nullable = false)))
+
+    val df = load(
+      source = classOf[SimpleFSBasedSource].getCanonicalName,
+      options = Map(
+        "path" -> basePath.toString,
+        "schema" -> dataSchema.json))
+
+    df.queryExecution.analyzed.collect {
+      case LogicalRelation(relation: SimpleFSBasedRelation) =>
+        assert(relation.dataSchema === dataSchema)
+      case _ =>
+        fail("Couldn't find expected SimpleFSBasedRelation instance")
+    }
+
+    val expectedSchema =
+      StructType(
+        Seq(
+          StructField("a", IntegerType, nullable = false),
+          StructField("p1", IntegerType, nullable = true),
+          StructField("b", StringType, nullable = false),
+          StructField("p2", StringType, nullable = true)))
+
+    assert(df.schema === expectedSchema)
+
+    df.select("b").where($"a" > 0 && $"p1" === 1).collect()
+
+    // Check for column pruning, filter push-down, and partition pruning
+    assert(TestResult.requiredColumns.toSet === Set("a", "b"))
+    assert(TestResult.filters === Seq(GreaterThan("a", 0)))
+    assert(TestResult.inputPaths === Seq(new Path(basePath, "p1=1").toString))
+  }
+
+  ignore("save() - partitioned table - Overwrite") {
+    testDF.save(
+      source = classOf[SimpleFSBasedSource].getCanonicalName,
+      mode = SaveMode.Overwrite,
+      options = Map(
+        "path" -> basePath.toString,
+        "schema" -> dataSchema.json),
+      partitionColumns = Seq("p1"))
+
+    // Written rows shouldn't contain dynamic partition column
+    val expectedRows = for (i <- 1 to 3; _ <- 1 to 2) yield Row(i, s"val_$i")
+    assert(TestResult.writtenRows.sameElements(expectedRows))
+  }
+
+  ignore("save() - partitioned table - Overwrite - select and overwrite the same table") {
+    TestResult.rowsToRead = testDF.rdd
+
+    val df = load(
+      source = classOf[SimpleFSBasedSource].getCanonicalName,
+      options = Map(
+        "path" -> basePath.toString,
+        "schema" -> dataSchema.json))
+
+    df.save(
+      source = classOf[SimpleFSBasedSource].getCanonicalName,
+      mode = SaveMode.Overwrite,
+      options = Map(
+        "path" -> basePath.toString,
+        "schema" -> dataSchema.json),
+      partitionColumns = Seq("p1"))
+
+    // Written rows shouldn't contain dynamic partition column
+    val expectedRows = for (i <- 1 to 3; _ <- 1 to 2) yield Row(i, s"val_$i")
+    assert(TestResult.writtenRows.sameElements(expectedRows))
+  }
+
+  ignore("save() - partitioned table - Append") {
```

Comment:
For append, we can have a little bit more tests. Here are two in my mind:

* having a df that only will be written to a subset of partitions;
* having a df that contains new partitioning value (we need to create new dirs).